src/Microsoft.AspNetCore.Server.Kestrel/Http/Frame.cs

```diff
@@ -851,7 +859,10 @@ public bool TakeMessageHeaders(SocketInput input, FrameRequestHeaders requestHea
                 while (!scan.IsEnd)
                 {
                     var beginName = scan;
-                    scan.Seek(ref _vectorColons, ref _vectorCRs);
+                    if (scan.Seek(ref _vectorColons, ref _vectorCRs) == -1)
+                    {
+                        return false;
```

Comment:
> If you seek when there's no input available, the iterator doesn't move forward.

I don't think this is accurate if I'm understanding what you're saying. The way `Seek` reads to me is that it advances the iterator to either the first character that matches or the end of the end of the input. `_block` and `_index` are always updated before returning even when it returns -1 (unless `IsDefault` is true which it never would be).

I do think there could be an issue if more input is added as the `TakeStartLine` or `TakeMessageHeaders` methods are running. In this way `Seek` *could* return -1 while the following `Take()` might return an actual character. This breaks the assumptions that these methods make. Namely that if the complete input for the start line or a given header isn't available fore the entire duration of the method, the method should return false.

It should be safe to return false even if all the incoming data hasn't been processed, since if the `examined` argument to `ConsumingComplete` doesn't point to the end of the input, the next `await SocketInput` should complete immediately, and the `TakeStartLine/MessageHeaders` method should be called again.

So in the end, I'm now convinced that we should always check the return value of `Seek` and return false if it's -1, but not because it's not advancing the iterator.